app/src/main/java/com/navapp/navigation/route/RouteHandler.java

```diff
@@ -0,0 +1,45 @@
+package com.navapp.navigation.route;
+
+import com.navapp.navigation.destination.Destination;
+import com.navapp.navigation.destination.data.Address;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+public class RouteHandler {
+    private final Route route;
+    private final Deque<Destination> remaining;
+
+    public RouteHandler(Route route) {
+        this(route, new ArrayDeque<>(route.getDestinations()));
+    }
+
+    public RouteHandler(Route route, Deque<Destination> remaining) {
+        this.route = route;
+        this.remaining = remaining;
+    }
+
+    public RouteHandler addDestination(Address currentAddress, Destination newDestination) {
+        remaining.add(newDestination);
+
+        return new RouteHandler(route.addDestination(newDestination), RouteOptimizer.optimizeRoute(currentAddress, remaining, route.getEnd()));
```

Comment:
This class is mutable, so why create a new `RouteHandler`?

Reply:
But it is not mutable